07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+				imageLayoutTransitionBarrier1.newLayout = asset::IImage::LAYOUT::GENERAL; // TODO: use more suitable layout
+				imageLayoutTransitionBarrier1.image = images[imageIdx].get();
+				imageLayoutTransitionBarrier1.subresourceRange = imgSubresourceRange;
+			}
+
+			core::smart_refctd_ptr<ISemaphore> imgFillSemaphore = m_device->createSemaphore(0); // TODO: don't create semaphore every iteration
+			IQueue::SSubmitInfo::SCommandBufferInfo imgFillCmdBuffInfo = { cmdBuff.get() };
+
+			IQueue::SSubmitInfo::SSemaphoreInfo imgFillSemaphoreWaitInfo = {
+				.semaphore = imgFillSemaphore.get(),
+				.value = 1,
+				.stageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS
+			};
+
+			IQueue::SSubmitInfo imgFillSubmitInfo = {
+				.waitSemaphores = {&imgFillSemaphoreWaitInfo, 1},
+				.commandBuffers = {&imgFillCmdBuffInfo, 1}
+			};
+
+			transferUpQueue->submit({ &imgFillSubmitInfo, 1 });
+
+			IQueue::SSubmitInfo::SSemaphoreInfo imgFillSemaphoreInfo =
+			{
+				.semaphore = imgFillSemaphore.get(),
+				.value = 0,
+				.stageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS
+			};
+			SIntendedSubmitInfo intendedSubmit = {
+				.frontHalf = {.queue = transferUpQueue, .waitSemaphores = {}, .commandBuffers = {&imgFillCmdBuffInfo, 1}}, .signalSemaphores = {&imgFillSemaphoreInfo, 1}
+			};
+			
+			cmdBuff->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo pplnBarrierDepInfo0;
+			pplnBarrierDepInfo0.imgBarriers = std::span(&imageLayoutTransitionBarrier0, &imageLayoutTransitionBarrier0 + 1);
+			if (!cmdBuff->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, pplnBarrierDepInfo0))
+				logFailAndTerminate("Failed to issue barrier!\n");
+
+			transferUpQueue->startCapture();
+			bool uploaded = m_utils->updateImageViaStagingBuffer(
+				intendedSubmit, cpuImages[imageIdx]->getBuffer(), cpuImages[imageIdx]->getCreationParameters().format,
+				images[imageIdx].get(), IImage::LAYOUT::TRANSFER_DST_OPTIMAL, cpuImages[imageIdx]->getRegions()
+			);
+			if (!uploaded)
+				logFailAndTerminate("Couldn't update image data.\n");
+
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo pplnBarrierDepInfo1;
+			pplnBarrierDepInfo1.imgBarriers = std::span(&imageLayoutTransitionBarrier1, &imageLayoutTransitionBarrier1 + 1);
+
+			if(!cmdBuff->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, pplnBarrierDepInfo1))
+				logFailAndTerminate("Failed to issue barrier!\n");
+
+			cmdBuff->end();
+
+			IQueue::SSubmitInfo submitInfo[1];
+			IQueue::SSubmitInfo::SCommandBufferInfo cmdBuffSubmitInfo[] = { {cmdBuff.get()} };
+			IQueue::SSubmitInfo::SSemaphoreInfo signalSemaphoreSubmitInfo[] = { { .semaphore = m_imagesLoadedSemaphore.get(), .value = imageIdx+1, .stageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS } };
+			submitInfo[0].commandBuffers = cmdBuffSubmitInfo;
+			submitInfo[0].signalSemaphores = signalSemaphoreSubmitInfo;
+			getTransferUpQueue()->submit(submitInfo);
```

Comment:
here's the thing, to use `SIntededSubmit` properly you need to cast (implicit convert) it to `SSubmitInfo`

then you can mofidy the signal span slightly to also signal the `m_imagesLoadedSemaphore`

Reply:
The `m_imagesLoadedSemaphore` you signal from and wait at ALL (because of the layout transition)
and `imgFillSemaphore` you signal from and wait at TRANSFER